File: tools/ez_tools/prefetch_parser.py
import os
import pandas as pd
from utils.discovery import find_artifact_files, load_csv_with_progress
from collector.collector import add_rows
from utils.logger import print_and_log

def process_prefetch(ez_dir: str, batch_size: int, base_dir: str):
    artifact_name = "Prefetch"
    print_and_log(f"[{artifact_name}] Scanning for relevant CSVs under: {ez_dir}")

    prefetch_files = find_artifact_files(ez_dir, base_dir, artifact_name)

    if not prefetch_files:
        print_and_log(f"[{artifact_name}] No Prefetch files found.")
        return

    date_columns = [
        ("SourceCreated", "Source Created"),
        ("SourceModified", "Source Modified"),
        ("SourceAccessed", "Source Accessed"),
        ("LastRun", "Last Run"),
        ("Volume0Created", "Volume Created"),
    ]
    for i in range(7):
        date_columns.append((f"PreviousRun{i}", f"Previous Run {i}"))

    for file_path in prefetch_files:
        print_and_log(f"[{artifact_name}] Processing: {file_path}")
        total_rows = 0
        try:
            for df in load_csv_with_progress(file_path, batch_size, artifact_name=artifact_name):
                timeline_data = []

                for _, row in df.iterrows():
                    for col, label in date_columns:
                        timestamp = row.get(col, "")
                        dt = pd.to_datetime(timestamp, utc=True, errors="coerce")
                        if pd.isnull(dt):
                            continue

                        dt_str = dt.isoformat().replace("+00:00", "Z")

                        timeline_row = {
                            "DateTime": dt_str,
                            "TimestampInfo": label,
                            "ArtifactName": artifact_name,
                            "Tool": "EZ Tools",
                            "Description": "Program Execution",
                            "DataPath": row.get("SourceFilename", ""),
                            "DataDetails": row.get("ExecutableName", ""),
                            "Count": row.get("RunCount", ""),
                            "EvidencePath": os.path.relpath(file_path, base_dir) if base_dir else file_path,
                        }

                        timeline_data.append(timeline_row)

                total_rows += len(timeline_data)
                add_rows(timeline_data)
        except Exception as e:
            print_and_log(f"[{artifact_name}] Failed to parse {file_path}: {e}")
            continue

        print_and_log(f"[✓] Parsed {total_rows} timeline rows from: {file_path}")

File: ForensicTimeliner.Core/Tools/EZTools/PrefetchParser.cs
using CsvHelper;
using CsvHelper.Configuration;
using ForensicTimeliner.Models;
using ForensicTimeliner.Interfaces;
using ForensicTimeliner.Models;
using ForensicTimeliner.Utils;
using Spectre.Console;
using System.Globalization;
using System.IO;
using System.Text.RegularExpressions; // Make sure this is included

namespace ForensicTimeliner.Tools.EZTools;

public class PrefetchParser : IArtifactParser
{
    public List<TimelineRow> Parse(string inputDir, string baseDir, ArtifactDefinition artifact, ParsedArgs args)
    {
        var rows = new List<TimelineRow>();

        Logger.PrintAndLog($"[>] - [{artifact.Artifact}] Scanning for relevant CSVs under: [{inputDir}]", "SCAN");

        var files = Discovery.FindArtifactFiles(inputDir, baseDir, artifact.Artifact);
        if (!files.Any())
        {
            Logger.PrintAndLog($"[!] - [{artifact.Artifact}] No matching files found in: {inputDir}", "WARN");
            return rows;
        }

        // Regex pattern to match the volume prefix
        var volumePattern = @"\\VOLUME\{[^}]+\}\\";

        foreach (var file in files)
        {
            int timelineCount = 0;
            Logger.PrintAndLog($"[+] - [{artifact.Artifact}] Processing: {Path.GetRelativePath(baseDir, file)}", "PROCESS");

            try
            {
                using var reader = new StreamReader(file);
                using var csv = new CsvReader(reader, new CsvConfiguration(CultureInfo.InvariantCulture)
                {
                    HeaderValidated = null,
                    MissingFieldFound = null
                });

                var records = csv.GetRecords<dynamic>();
                foreach (var record in records)
                {
                    var dict = (IDictionary<string, object>)record;

                    var parsedDt = dict.GetDateTime("RunTime");
                    if (parsedDt == null) continue;

                    string dtStr = parsedDt.Value.ToString("o").Replace("+00:00", "Z");
                    string exe = dict.GetString("ExecutableName");

                    // Extract just the filename from the full path
                    string exeFileName = Path.GetFileName(exe);

                    // Remove the volume prefix pattern using regex
                    string cleanPath = Regex.Replace(exe, volumePattern, "", RegexOptions.IgnoreCase);

                    rows.Add(new TimelineRow
                    {
                        DateTime = dtStr,            // RunTime mapped to DateTime
                        TimestampInfo = "Run Time",
                        ArtifactName = artifact.Artifact,
                        Tool = artifact.Tool,
                        Description = artifact.Description,
                        DataPath = cleanPath,        // Path without volume prefix
                        DataDetails = exeFileName,   // Just the filename
                        EvidencePath = Path.GetRelativePath(baseDir, file)
                    });

                    timelineCount++;
                }

                Logger.PrintAndLog($"[✓] - [{artifact.Artifact}] Parsed {timelineCount} timeline rows from: {Path.GetFileName(file)}", "SUCCESS");
                LoggerSummary.TrackSummary(artifact.Tool, artifact.Artifact, timelineCount);
            }
            catch (Exception ex)
            {
                Logger.PrintAndLog($"[{artifact.Artifact}] Failed to parse {file}: {ex.Message}", "ERROR");
            }
        }

        return rows;
    }
}